Windows Installer: Upgrade to WiX Toolset 5

jkarasti opened the pull request:
This upgrades tooling used to build the Dangerzone.msi installer to WiX Toolset 5.

The Major change in this, apart from what running `wix convert` on the old generated WiX authoring, is that the target directory for each component gets set in the `Directory` attribute of the component instead of a separate `Directory` element one level above in the XML hierarchy. All components get grouped under a `ComponentGroup` that gets referenced in the `Feature` tree, which makes things much simpler aswell.

There are still a couple of open questions:

#### Should `AllowSameVersionUpgrades="yes"` be removed?

According to the [documentation](https://wixtoolset.org/docs/schema/wxs/majorupgrade/#attributes) setting it useful if the installer uses a fourth version number. But Dangerzone doesn't use one so it should be save to remove it. Running `wix msi validate` on the built msi also prints a warning about this.

#### What to set the value of  `InstallerVersion` to?

In this its set to `200` to make 64 bit installer work, but it could also be removed or set to the default  `500`, which would cap the lowest version of windows Dangerzone supports to Windows 7.

#### Should Dangerzone also set a desktop icon?

This is also a possibility, but opinions vary is this is good practice or not. Also this should come with an option to not install the icon, but that would require developing some custom dialogs for the installer.

#### Customise the installer theming more?

This should probably be its own issue. The installer, while limited, allows for a bit customization, mainly by addind a banner image that would replace the red cd icon shown during install currently. The installer already uses a custom image shown at the start of the installation . For more, see:

https://wixtoolset.org/docs/tools/wixext/wixui/#replacing-the-default-bitmaps
https://stackoverflow.com/questions/48713061/how-do-you-change-the-red-cd-icon-on-wix

> [!CAUTION]
> One unfortunate side effect this has is that installing the newer installer does not uninstall older versions of Dangerzone.
> The root cause for this is a changed default for the `Scope` attribute in the `Package` element. In WiX 3 it was apparently
> set to `perUser` by mistake and in WiX 4 its set to `perMachine`

One workaround for the uninstallation issue would be to set `Scope="perUser"` attribute in the `Package` element. But MSIs apparently have poor support for per user packages, though I havent figured out how exactly just yet.

Another solution could be to bite the bullet this once and guide users on Windows into uninstalling the older version of Dangerzone somehow, for example a blog post or a pinned issue, or maybe within Dangerzone itself. SInce this pull request also makes Dangerzone a proper 64 bit installation, the new version should install itself alongside the old one just fine without breaking anything. Though it'll show up twice in installed programs and the start menu, leaving it up to the user to discover older version is still installed and uninstall it. Not exactly the best user experience.

closes #602

apyrgio commented on install/windows/build-wxs.py:
```diff
@@ -123,122 +84,160 @@ def main():
+    wix_el = ET.Element(
+        "Wix",
+        {
+            "xmlns": "http://wixtoolset.org/schemas/v4/wxs",
```
Just wanted to say that while this is not supposed to be a link, it redirects you to Smashing Pumpkin's "Bullet with Butterfly Wings", and I'm down with it.

jkarasti commented on .github/workflows/ci.yml:
```diff
@@ -21,10 +21,14 @@ jobs:
+        with:
+          dotnet-version: "8.x"
+      - name: Install WiX Toolset
+        run: dotnet tool install --global wix --version 5.0.1
```
We could set the version in an environment variable would that be enough?

```yaml
  env:
    wix_version: 5.0.1
- name: Install WiX Toolset
   run: dotnet tool install --global wix --version $wix_version
- name: Add WiX UI extension
   run: wix extension add --global WixToolset.UI.wixext/$wix_version
```

apyrgio replied:
It's not the duplication that worries me, but whether a future 5.1 Wix release breaks something. I'll give it a shot locally.

jkarasti replied:
Oops, yes, I started looking into the wrong thing. We could set the WiX version with a matrix stradegy. Something like:

```yaml
strategy:
      matrix:
        wix_version: [5.0.0, 5.0.1]

- name: Install WiX Toolset
   run: dotnet tool install --global wix --version ${{ matrix.wix_version }}
- name: Add WiX UI extension
   run: wix extension add --global WixToolset.UI.wixext/${{ matrix.wix_version }}
```

apyrgio replied:
Actually, I was thinking of something like the following:

```
dotnet tool install --global wix --version 5.*
wix extension add --global WixToolset.UI.wixext/5
```

I just realized though that there's an open issue for the latter, and it won't fetch the latest 5.x.y release of WixToolset.UI.wixext (see https://github.com/wixtoolset/issues/issues/8033)

Anyway, we can pin the extension version for now, but my suggestion is to always install the latest WiX version using `dotnet tool install`, if you agree.

jkarasti replied:
Right, now I got it. I hardcoded the version to avoid a situation where the windows installer accidentally gets built with a newer version of WiX Toolset and breaks in unexpected ways. E.g See the whole Scope debacle. 

We could drop the version numbers in CI to catch any regressions, but keep them in the build instructions to prevent accidentally upgrading WiX, fox example. I'm not particularly opinionated on this.

jkarasti replied:
And on that note, should I also remove the dotnet version? The setup-dotnet action picks whatever is the newest version installed in the runner if its not defined. I picked 8 since that's the version you'll get offered when downloading .NET SDK.

apyrgio replied:
> Or maybe even have two jobs in CI. One installs WiX without a version number to test for regressions and the other uses whichever version is used to build the current released installer and gets bumped when the installer is upgraded to a newer version of WiX

Excellent, that's the best of both worlds. Our CI is undergoing a major change, so no need to do this in this PR. I'll create a separate issue to track it.

apyrgio commented:
Amazing work here @jkarasti. Not only did you manage to update our scripts to use WiX 5, but you broke down the changes in 20 (!) short and sweet commits. The changes make sense, and I'm testing them in a Windows box now.

Now, to your questions, here's my understanding:

## Should `AllowSameVersionUpgrades="yes"` be removed?

My understanding is that previously, we specified `*` as the product code:

https://github.com/freedomofpress/dangerzone/blob/791444cd5dd983f87696c4d16313f2d3a6af515d/install/windows/build-wxs.py#L167

With this PR, we don't specify something, which means that the default behavior will trigger:

> By default, a new ProductCode property will be generated with every build enabling major upgrades.

_(from https://wixtoolset.org/docs/schema/wxs/package/#attributes)_

So, if we remove `AllowSameVersionUpgrades="yes"`, this means that we cannot simply rebuild the Dangerzone MSI, if Wix announces a CVE for instance (this has [happened](https://github.com/freedomofpress/dangerzone/releases/tag/v0.6.0) in the past btw). We will need to also bump the patch version of Dangerzone. This is not bad per se, but it's a limitation us devs need to be aware of. 

## What to set the value of `InstallerVersion` to?

Setting it to 200 - or even removing it, since it's the default - is fine. After all, Docker Desktop (well, WSL2) works from Windows 10 onwards, so we're capped by this requirement.

## Should Dangerzone also set a desktop icon?

Not pressing, looks like a future improvement to me.

## Customise the installer theming more?

Another future improvement, but thanks for the context. Once we merge the PR, we should refer to your links in a separate issue, as you have correctly mentioned.

apyrgio commented on BUILD.md:
````diff
@@ -433,9 +433,14 @@ poetry shell
-* Go to https://dotnet.microsoft.com/download/dotnet-framework and download and install .NET Framework 3.5 SP1 Runtime. I downloaded `dotnetfx35.exe`.
-* Go to https://wixtoolset.org/releases/ and download and install WiX toolset. I downloaded `wix314.exe`.
-* Add `C:\Program Files (x86)\WiX Toolset v3.14\bin` to the path ([instructions](https://web.archive.org/web/20230221104142/https://windowsloop.com/how-to-add-to-windows-path/)).
+Install [.NET SDK](https://dotnet.microsoft.com/en-us/download) version 6 or later.
+Install [WiX toolset](https://wixtoolset.org/) v5 and add the WiX UI extension.
+```
+dotnet tool install --global wix --version 5.0.1
+wix extension add --global WixToolset.UI.wixext/5.0.1
+```
+> [!IMPORTANT]  
+> To prevent compatibility issues, ensure that all installed WiX plugins match at least the major version of the Wix Toolset.
````
I had to reopen the Windows shell during the installation of WiX. Let's reflect this in our notes:. Proposed change:

Install [.NET SDK](https://dotnet.microsoft.com/en-us/download) version 6 or later. Then, open a terminal and install [WiX toolset](https://wixtoolset.org/) v5:

```
dotnet tool install --global wix --version 5.*
```

Install the WiX UI extension in a **new terminal**, in order to use the newly installed `wix` tool:

```
wix extension add --global WixToolset.UI.wixext/5.0.1
```

> [!IMPORTANT]  
> To prevent compatibility issues, ensure that all installed WiX plugins match at least the major version of the Wix Toolset.

jkarasti replied:
That's odd. I could install both wix and the ui extension without opening a new terminal. Though I think agree we should mention opening the terminal explicitly

apyrgio replied:
I'm 99% sure, I also a message that said I needed to open a new terminal to refresh the path. I see a [tutorial](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-use) mentioning this as well:

> If this command fails, you may need to open a new terminal to refresh the PATH.

apyrgio commented:
FYI, I tested your branch and it works great. I can verify that the previous Dangerzone installation is not removed, which is a major bummer. I'll dig more around to figure out if we can do something about that.

apyrgio commented:
I tested with `Scope="perUser"` and I get the following error:

![image](https://github.com/user-attachments/assets/411290e2-301f-4d0e-8c82-17fa782c5d3e)

This is similar to this [SO post](https://stackoverflow.com/questions/77839235/wix-how-to-upgrade-per-user-installation-including-services), in which they had to do some shenanigans to elevate their privileges on installation time.

I wonder, is there a way to detect that a similar package exists, and ask users to remove it from "Programs and Features", like we do in the `MajorUpgrade` element?

jkarasti commented:
> I tested with `Scope="perUser"` and I get the following error:
> 
> ![image](https://private-user-images.githubusercontent.com/2507626/370732678-411290e2-301f-4d0e-8c82-17fa782c5d3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyNzYwNDksIm5iZiI6MTcyNzI3NTc0OSwicGF0aCI6Ii8yNTA3NjI2LzM3MDczMjY3OC00MTEyOTBlMi0zMDFmLTRkMGUtOGM4Mi0xN2ZhNzgyYzVkM2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjVUMTQ0OTA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjYzMjQ0ZTQwMGM1NmU5MmMzZGRkYThkODc0YWZhNzU4NDVkOGFmZDgyNmNhMjFkYmVjODQ5YWU1NTA0ZjAxYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.UhOWYn7Wj2jA5xEPcw3RGaKo0QmFie8PvV1zqJUl8ic)
> 
> This is similar to this [SO post](https://stackoverflow.com/questions/77839235/wix-how-to-upgrade-per-user-installation-including-services), in which they had to do some shenanigans to elevate their privileges on installation time.
> 
> I wonder, is there a way to detect that a similar package exists, and ask users to remove it from "Programs and Features", like we do in the `MajorUpgrade` element?

I found [this](https://www.firegiant.com/docs/wix/v3/tutorial/upgrades-and-modularization/checking-for-oldies/). maybe we could add some custom action to detect and block the installation, I haven't looked into this much yet.

jkarasti commented:
> ## Should `AllowSameVersionUpgrades="yes"` be removed?
> 
> My understanding is that previously, we specified `*` as the product code:
> 
> https://github.com/freedomofpress/dangerzone/blob/791444cd5dd983f87696c4d16313f2d3a6af515d/install/windows/build-wxs.py#L167
> 
> With this PR, we don't specify something, which means that the default behavior will trigger:
> 
> > By default, a new ProductCode property will be generated with every build enabling major upgrades.
> 
> _(from https://wixtoolset.org/docs/schema/wxs/package/#attributes)_

Just to note: Setting the value of `Id` to `*` means WiX will generate a GUID automatically, so this is doesen't introduce a change in behaviour. See docs for WiX 3 [here](https://wixtoolset.org/docs/v3/xsd/wix/product/) and [here](https://wixtoolset.org/docs/v3/xsd/wix/simple_type_autogenguid/)

> So, if we remove `AllowSameVersionUpgrades="yes"`, this means that we cannot simply rebuild the Dangerzone MSI, if Wix announces a CVE for instance (this has [happened](https://github.com/freedomofpress/dangerzone/releases/tag/v0.6.0) in the past btw). We will need to also bump the patch version of Dangerzone. This is not bad per se, but it's a limitation us devs need to be aware of.

Ah, I didn't consider needing to rebuild the installer with a same version number. The main motivation for removing it is to make validation pass without warnings.

By the way, looks like making `AllowSameVersionUpgrades="yes"` the default was considered, but doing it would sst off the warning by default and disabling the warning would disable many other checks aswell. See [this episode](https://youtu.be/vqiEVfeDjpw?t=1291) of Rob Menschings excellent series on WiX Toolset 4. (This was a great resource with this upgrade. btw.)

Also. WiX Toolset 5 introduces a brand new feature: [Files](https://wixtoolset.org/docs/fivefour/#file-harvesting). I'm working on a new version of this PR that integrates it int the .wxs build script. In a nutshell, the entire file harversting dance in `build-wxs.py` (and ~3000 lines of code in the generated Dangerzone.wxs) can be replaced with a sinle line:

```
<Files Include="exe.win-amd64-3.12\**" />
```

apyrgio commented on install/windows/build-wxs.py:
I'm _really_ excited to test this hack out. Thanks a lot!

jkarasti replied:
I tested it some more and, unfortunately, it get's us about 99% of the way there. With this workaround, the installer can find and uninstall the older version of Dangerzone and the subsequent upgrades work as well. But after reverting the hack the duplicate version issue returns, even if the older installer was built with WiX v5 (!).

apyrgio commented:
Hey @jkarasti. Just wanted to let you know that we have wrapped Dangerzone 0.8.0. If there's something I can take a look at or help you with, now is the time to let me know!

jkarasti commented:
> Hey @jkarasti. Just wanted to let you know that we have wrapped Dangerzone 0.8.0. If there's something I can take a look at or help you with, now is the time to let me know!

Hi. I'd say the biggest TODO on here is figuring out the better way around the changed install scope, since the one I have here currently seems to break the installer in some way.

I've been experimenting with using [this](https://wixtoolset.org/docs/v3/howtos/files_and_registry/check_the_version_number/) trick to try detecting the version of currently installed Dangerzone and then stopping the install using the Condition attribute in a WiX Launch element. I've almost got it working, but it still needs work. A bit annoyingly, since #909 got included in 0.8.0, this way now has an edge case where it wont work if the user chooses a non-default installation location.

An another way I've been trying a bit would be running a PowerShell script to parse through the windows registry for an existing install. I have a script that should be able to detect Dangerzone reliably regardless of where its installed. I should be able to include the script using a WiX Binary element and then run it using a Custom Action, but still need to figure out a way to actually do that. The issue I have with this approach is that snooping through what the user has installed on their computer doesen't feel particularly security minded. At the moment the registry keys for Dangerzone live in ´HKLM/Software/Wow6432Node/Microsoft/Windows/CurrentVersion/Uninstall/´ under some GUID. One thought I had while writing this would be to see if the GUID is the same for some version of Dangerzone installed and try detecting those instead.

Other than that, I'm fairly confident this is ready for review, except for maybe a bunch of bikeshedding over minor stuff like `AllowSameVersionUpgrades`, documentation and such.

apyrgio commented:
Great, thanks for the progress report. I'll give this PR another spin soon and review the latest additions. As for our main blocker, sigh, I'll try to take a look as well, but it may take me a while to build the context you have. I'll keep you posted...

apyrgio commented:
Just checked the diff between the state of the PR when I reviewed it and its current state. I'm fine with the changes, and the only thing remaining is indeed migrating the installation scope. Haven't tested your proposed changes on that part yet, since you mention some edge cases.

Regarding `AllowSameVersionUpgrades`, your explanation about being spec-compliant makes sense, so I'm fine with removing it. As for the documentation fixes, they look ok to me.

apyrgio commented:
Out of curiosity, I see that we can trigger the uninstall dialog via:

```
MsiExec.exe /X{GUID}
```

(taken from this [SO post](https://stackoverflow.com/a/19807786), not tested)

I suppose we can trigger this command during the installation of our .MSI. Any downside to that?

---

Btw, if it takes too much of our energy to fix this issue, then that's fine. You mentioned already that:

> 

Another solution could be to bite the bullet this once and guide users on Windows into uninstalling the older version of Dangerzone somehow, for example a blog post or a pinned issue, or maybe within Dangerzone itself. SInce this pull request also makes Dangerzone a proper 64 bit installation, the new version should install itself alongside the old one just fine without breaking anything. Though it'll show up twice in installed programs and the start menu, leaving it up to the user to discover older version is still installed and uninstall it. Not exactly the best user experience.

I would be fine with adding a download link that says "Read **this** if you have installed Dangerzone version v0.8.0 or older", and explain to the user that they need to uninstall their previous Dangerzone version first. It's not the best thing ever, but we're reaching Wix v3 deprecation soon, so that's the hand we're dealt with.

jkarasti commented:
> Just checked the diff between the state of the PR when I reviewed it and its current state. I'm fine with the changes, and the only thing remaining is indeed migrating the installation scope. Haven't tested your proposed changes on that part yet, since you mention some edge cases.

I would recommend testing the installer in a vm or a [Windows Sandbox](https://learn.microsoft.com/en-us/windows/security/application-security/application-isolation/windows-sandbox/windows-sandbox-overview) As it turns out, if you're clever enough, you can tell WiX to build an installer which will launch if and only if it's not installed... that's apparently true when uninstalling it aswell. Fun times...

Which brings us to:

> Btw, if it takes too much of our energy to fix this issue, then that's fine. You mentioned already that:

> Another solution could be to bite the bullet this once and guide users on Windows into uninstalling the older version of Dangerzone somehow, for example a blog post or a pinned issue, or maybe within Dangerzone itself. SInce this pull request also makes Dangerzone a proper 64 bit installation, the new version should install itself alongside the old one just fine without breaking anything. Though it'll show up twice in installed programs and the start menu, leaving it up to the user to discover older version is still installed and uninstall it. Not exactly the best user experience.

> I would be fine with adding a download link that says "Read this if you have installed Dangerzone version v0.8.0 or older", and explain to the user that they need to uninstall their previous Dangerzone version first. It's not the best thing ever, but we're reaching Wix v3 deprecation soon, so that's the hand we're dealt with.

I'm leaning towards agreeing as of now, maybe it would be enough to mention uninstalling the older version of Dangerzone in the release notes or blog post with some degree of urgency and maybe add an announcement  banner on the website like was done in freedomofpress/dangerzone.rocks/#51. Though I do have a couple of ideas for a workaround, for example, adding a registry key, something like `NewInstaller=1` and then using that as a marker for showing an error message when launching the installer. But this could be a follow up.

> Regarding `AllowSameVersionUpgrades`, your explanation about being spec-compliant makes sense, so I'm fine with removing it. As for the documentation fixes, they look ok to me.

I removed it.

> Out of curiosity, I see that we can trigger the uninstall dialog via:
> 
> ```
> MsiExec.exe /X{GUID}
> ```
> 
> (taken from this [SO post](https://stackoverflow.com/a/19807786), not tested)
> 
> I suppose we can trigger this command during the installation of our .MSI. Any downside to that?
> 

I think this could work, maybe use a WiX `Custom Action` Element to run it, though we'd need to know which GUIDs to look for first. I suspect it's whatever WiX assingns to the `ProductCode` property in the Package element.

I pushed a couple small tweaks for the build docs in a separate fixup commit, please check I didn't make it too complicated.

And I also  dropped the hardcoded versions with the msi build in CI. Thinking about it, having two separate actions for it might be a bit overkill, especially since there shouldn't be any meaningful difference between them before WiX Toolset v6 is released ([rc.1](https://github.com/wixtoolset/issues/milestone/27) has a release date in February). I think the idea is releasing more frequent major versions of WiX with less breaking changes going foward, so upgrading to WiX v6 should be (fingers crossed) be much less painfull aswell.

apyrgio commented:
Tested the installer and it works, with the known caveat. We have to discuss internally whether it's fine to proceed as is and notify users in our installation page, or whether we need to pursue a proper fix. I'll let you know next week what's the case.

Note that I see in the code the workaround you were experimenting with. In my case, it didn't uninstall the 0.8.0 Dangerzone installation, and I've been using the default installation path. Just letting you know in case it's useful.

jkarasti commented:
Trying to avoid leaving the Scope issue as is, I took an another deep dive to WiX Toolset `Launch` and `Custom Action` elements. Unfortunately there really doesen't seem to be a "works like a charm" solution here, at least I haven't found it yet, but I did manage to come up with a somewhat better fix which won't result in a somehow broken installer in `[jkarasti:602-upgrade-wix-toolset-fix-regsearch](https://github.com/jkarasti/dangerzone/tree/602-upgrade-wix-toolset-fix-regsearch)`. Here I'm using a registry search to look for Istallation of Dangerzone 0.8.0 and stop the installation from starting with a WiX `Launch` Element. It's not quite perfect. For one, it relies on the ProductId of 0.8.0 installation being `{03C2D2B2-9955-4AED-831F-DA4E67FC0FDB}` though I think that should always be the case. And second, most likely thanks to the different install scope, you can install 0.8.0 after installing Dangerzone using the new installer, but cannot uninstall the new installation before uninstalling 0.8.0 with this fix included. Note that this is still very much experimental, so test it in a VM.

I also tried using MsiExec.exe to trigger an uninstall during the install but that didn't work at all. Or it did insofar as trying to run MsiExec.exe during the install, but I completely forgot [you can't run two msi at the same time](https://stackoverflow.com/a/49503544). But from that SO answer, it could maybe be possible if we converted the installer into a WiX Bundle which can run msi installers in a sequence. Though that would mean redesigning the installer from scratch so, unless shipping a WiX Bundle has some other amazing muat-have benefits, I don't think going that route is worth it.

Also, I've moved the fix that I had in this branch into `[[jkarasti:602-upgrade-wix-toolset-v2](https://github.com/jkarasti/dangerzone/tree/602-upgrade-wix-toolset-fix-old](https://github.com/jkarasti/dangerzone/tree/602-upgrade-wix-toolset-fix-old)` since it appears to break the install scope so I don't think we should include it.

apyrgio commented:
Just tested [jkarasti:602-upgrade-wix-toolset-fix-regsearch](https://github.com/jkarasti/dangerzone/tree/602-upgrade-wix-toolset-fix-regsearch) and it detected the Dangerzone 0.8.0 installation. However, it doesn't detect previous installations, such as 0.6.0. Our users may skip 0.8.0 for whatever reasons, so we have to expect such a scenario :-/

I think it's fine to merge it as is, and think of some instructions that we want to provide to users, instead of just letting them download the .msi.

jkarasti commented:
> Just tested [jkarasti:602-upgrade-wix-toolset-fix-regsearch](https://github.com/jkarasti/dangerzone/tree/602-upgrade-wix-toolset-fix-regsearch) and it detected the Dangerzone 0.8.0 installation. However, it doesn't detect previous installations, such as 0.6.0. Our users may skip 0.8.0 for whatever reasons, so we have to expect such a scenario :-/
> 

That's working as expected and my initial plan was to figure out the GUIDs for some amount of older versions of Dangerzone and add checks for them, but I've come up with a better fix. Instead of looking at the registy, the installer now checks if `dangerzone.exe` exists in `C:\Program Files (x86)\Dangerzone`. This should cover every version except for 0.8.0 with a non-default install folder, so I've also kept the registry search from [jkarasti:602-upgrade-wix-toolset-fix-regsearch](https://github.com/jkarasti/dangerzone/tree/602-upgrade-wix-toolset-fix-regsearch). The installer should now stop with an error message prompting to uninstall if either of the two are true and continue otherwise.

apyrgio commented:
Thanks again for the contribution @jkarasti. I just tested it in a Windows machine and everything works. Awesome job :star_struck:
